src/btree/bt_vrfy.c

```diff
@@ -582,6 +583,105 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
     return (0);
 }
 
+/*
+ * __verify_btree_id_with_meta --
+ *     With a given btree id, iterate through the metafile table and check if the btree id exists in
+ *     any table. If exists, we place the table URI into a variable otherwise should return
+ *     WT_NOTFOUND.
+ */
+static int
+__verify_btree_id_with_meta(WT_SESSION_IMPL *session, uint32_t btree_id, const char **uri)
+{
+    WT_CONFIG_ITEM id;
+    WT_CURSOR *meta_cursor;
+    WT_DECL_RET;
+    const char *meta_value;
+
+    WT_ERR(__wt_metadata_cursor(session, &meta_cursor));
+    while ((ret = meta_cursor->next(meta_cursor)) == 0) {
+        WT_ERR(meta_cursor->get_value(meta_cursor, &meta_value));
+        if ((ret = __wt_config_getones(session, meta_value, "id", &id)) == 0) {
+            if (btree_id == __wt_strtouq(id.str, NULL, 10)) {
+                meta_cursor->get_key(meta_cursor, uri);
+                break;
+            }
+        }
+        WT_ERR_NOTFOUND_OK(ret);
+    }
+err:
+    WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
+    return (ret);
+}
+
+/*
+ * __wt_verify_history_store_tree --
+ *     Verify the history store tree, by iterating and checking all the keys in the history and that
+ *     it exists in both the meta file and the respective data store table. Skips duplicate keys by
+ *     storing the previous key.
+ */
+int
+__wt_verify_history_store_tree(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor, *data_cursor;
+    WT_DECL_RET;
+    WT_ITEM hs_key, prev_hs_key;
+    WT_TIME_PAIR hs_start, hs_stop;
+    uint32_t btree_id, session_flags, prev_btree_id;
+    int cmp;
+    const char *uri;
+    bool first;
+
+    session_flags = 0;
+    /* Dodging uninitialized build error. Don't trust this initial value. */
+    prev_btree_id = 0;
+    uri = NULL;
+    first = true;
+    WT_CLEAR(prev_hs_key);
+    WT_CLEAR(hs_key);
+
+    WT_ERR(__wt_hs_cursor(session, &session_flags));
+    cursor = session->hs_cursor;
+
+    while ((ret = cursor->next(cursor)) == 0) {
+        WT_ERR(cursor->get_key(cursor, &btree_id, &hs_key, &hs_start.timestamp, &hs_start.txnid,
+          &hs_stop.timestamp, &hs_stop.txnid));
+
+        /*
+         *  Keep track of the previous comparison. The history store is stored in order, so we can
+         *  avoid redundant comparisons. The first flag is used to avoid comparing against
+         *  uninitialized variables.
+         */
+        if (!first && prev_btree_id == btree_id) {
+            WT_ERR(__wt_compare(session, NULL, &hs_key, &prev_hs_key, &cmp));
+            if (cmp == 0)
+                continue;
+        }
+
+        first = false;
+
+        prev_hs_key.data = hs_key.data;
+        prev_hs_key.size = hs_key.size;
+        prev_btree_id = btree_id;
```

Comment:
This looks dangerous to me. Imagine hs_key has a chunk of memory and `hs_key.data` points into `hs_key.mem`. This line of code happens, then `prev_hs_key.data` points into `hs_key.mem`. Then, we do the `get_key` call at line 646, and `hs_key.mem` is free'd and reallocated, now `prev_hs_key.data` points into freed' memory. Then we call `__wt_compare` and core dump.

Is that incorrect?

If I'm correct, I think you want a pointer to a `WT_ITEM` that gets swapped and the contents of the `WT_ITEM` aren't touched.

You might want to use scratch buffer allocation as well.

Reply:
@keithbostic 
I agree that this code isn't safe but I don't understand the solution.
To my understanding, the problem is that `prev_hs_key` is pointing at a buffer owned by the cursor. Then when we call `next`, it will get freed from underneath us. So I would have thought the solution would to use `__wt_buf_set` to `memcpy` the contents to the `WT_ITEM` so it survives the next iteration.

Reply:
Oh, I get it. Yeah, agreed, something like:
```
        if (!WT_DATA_IN_ITEM(&cursor->value))
            WT_RET(__wt_buf_set((WT_SESSION_IMPL *)cursor->session, &cursor->value,
              cursor->value.data, cursor->value.size));
```
(from `cursor.i`)